#### Data Wrangling

-   Identify duplicate values in the dataset.

-   Remove duplicate values from the dataset.

-   Identify missing values in the dataset.

-   Impute the missing values in the dataset.

-   Normalize data in the dataset.


<hr>


Import pandas module.


In [1]:
import pandas as pd

Load the dataset into a dataframe.


In [2]:
df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DA0321EN-SkillsNetwork/LargeData/m1_survey_data.csv")

## Finding duplicates


In [5]:
# your code goes here
duplicate_rows = df.duplicated().sum()
print(f"There are {duplicate_rows} duplicate rows.")

There are 154 duplicate rows.


In [6]:
duplicate_count = df.duplicated(subset='Respondent', keep=False).sum()
print(duplicate_count)



247


## Removing duplicates


Remove the duplicate rows from the dataframe.


In [7]:
# your code goes here
df = df.drop_duplicates()

Verify if duplicates were actually dropped.


In [8]:
# your code goes here
duplicate_rows_after_dropping = df.duplicated().sum()
print(f"There are {duplicate_rows_after_dropping} duplicate rows after dropping.")

There are 0 duplicate rows after dropping.


In [9]:
#How many rows are in the dataset now? 
number_of_rows = df.shape[0]
print(number_of_rows)


11398


In [14]:
#finding the unique rows in the coloumn 'Respondent'.
unique_number_of_unique_values = df['Respondent'].nunique()
unique_number_of_unique_values

11398

## Finding Missing values


Find the missing values for all columns.


In [18]:
# your code goes here
total_missing_values = df.isnull().sum().sum()
print(f"There are {total_missing_values} missing values in total.")



There are 30967 missing values in total.


Find out how many rows are missing in the column 'WorkLoc'


In [20]:
# your code goes here
missing_values_in_column = df['WorkLoc'].isnull().sum()
print(f"There are {missing_values_in_column} missing values in the 'WorkLoc' column.")



There are 32 missing values in the 'WorkLoc' column.


In [22]:
#finding blank rows under column 'Edlevel'
number_of_blank_rows = df['EdLevel'].isnull().sum()
number_of_blank_rows

112

In [23]:
#finding the number of rows that are missing in column 'Country'
number_of_missing_rows = df['Country'].isnull().sum()
number_of_missing_rows

0

## Imputing missing values


Find the  value counts for the column WorkLoc.


In [24]:
# your code goes here
value_counts = df['WorkLoc'].value_counts()
print(value_counts)


Office                                            6806
Home                                              3589
Other place, such as a coworking space or cafe     971
Name: WorkLoc, dtype: int64


In [25]:
#Find the value counts for the column Employment
value_counts = df['Employment'].value_counts()
print(value_counts)



Employed full-time    10968
Employed part-time      430
Name: Employment, dtype: int64


In [26]:
#Find the value counts for the column UndergradMajor
value_counts = df['UndergradMajor'].value_counts()
print(value_counts)


Computer science, computer engineering, or software engineering          6953
Information systems, information technology, or system administration     794
Another engineering discipline (ex. civil, electrical, mechanical)        759
Web development or web design                                             410
A natural science (ex. biology, chemistry, physics)                       403
Mathematics or statistics                                                 372
A business discipline (ex. accounting, finance, marketing)                244
A social science (ex. anthropology, psychology, political science)        210
A humanities discipline (ex. literature, history, philosophy)             207
Fine arts or performing arts (ex. graphic design, music, studio art)      161
I never declared a major                                                  124
A health science (ex. nursing, pharmacy, radiology)                        24
Name: UndergradMajor, dtype: int64


Identify the value that is most frequent (majority) in the WorkLoc column.


#make a note of the majority value here, for future reference
### Office is the majority value with 6806

Impute (replace) all the empty rows in the column WorkLoc with the value that you have identified as majority.


In [32]:
# Identify the majority value in the column
majority_value = df['WorkLoc'].mode().iloc[0]

# Impute the missing values with the majority value
df['WorkLoc'].fillna(majority_value,inplace=True)

print(df['WorkLoc'])

0                                                  Home
1                                                Office
2                                                  Home
3                                                  Home
4        Other place, such as a coworking space or cafe
                              ...                      
11547                                              Home
11548                                              Home
11549                                            Office
11550                                              Home
11551                                            Office
Name: WorkLoc, Length: 11398, dtype: object


After imputation there should ideally not be any empty rows in the WorkLoc column.


Verify if imputing was successful.


In [14]:
missing_values_after_imputing = df['WorkLoc'].isnull().sum()
print(f"There are {missing_values_after_imputing} missing values in the 'WorkLoc' column after imputing.")


There are 0 missing values in the 'WorkLoc' column after imputing.


## Normalizing data


<hr>


List out the various categories in the column 'CompFreq'


In [15]:
unique_categories = df['CompFreq'].unique()
print(f"The unique categories in the 'CompFreq' column are: {unique_categories}")

The unique categories in the 'CompFreq' column are: ['Yearly' 'Monthly' 'Weekly' nan]


In [47]:
#drop the "nan" in CompFreq column
df = df.dropna(subset=['CompFreq'])

In [48]:
# how many respondents are getting paid yearly
category_counts = df['CompFreq'].value_counts()
print(category_counts)

Yearly     6073
Monthly    4788
Weekly      331
Name: CompFreq, dtype: int64


In [52]:
df['NormalizedAnnualCompensation'] = 0

for index, row in df.iterrows():
    if row['CompFreq'] == 'Yearly':
        df.at[index, 'NormalizedAnnualCompensation'] = row['CompTotal']
    elif row['CompFreq'] == 'Monthly':
        df.at[index, 'NormalizedAnnualCompensation'] = row['CompTotal'] * 12
    elif row['CompFreq'] == 'Weekly':
        df.at[index, 'NormalizedAnnualCompensation'] = row['CompTotal'] * 52

# calculate mean for the 'NormalizedAnnualCompensation' column. 
median_Normalized_Annual_Compensation = df[ 'NormalizedAnnualCompensation'].median()
print("Normalized Annual Compensation median:", median_Normalized_Annual_Compensation)

Normalized Annual Compensation median: 100000.0


In [53]:
df.head()

Respondent                      MainBranch Hobbyist  \
0           4  I am a developer by profession       No   
1           9  I am a developer by profession      Yes   
2          13  I am a developer by profession      Yes   
3          16  I am a developer by profession      Yes   
4          17  I am a developer by profession      Yes   

                                         OpenSourcer  \
0                                              Never   
1                         Once a month or more often   
2  Less than once a month but more than once per ...   
3                                              Never   
4  Less than once a month but more than once per ...   

                                          OpenSource          Employment  \
0  The quality of OSS and closed source software ...  Employed full-time   
1  The quality of OSS and closed source software ...  Employed full-time   
2  OSS is, on average, of HIGHER quality than pro...  Employed full-time   
3  The quality of OSS and closed source software ...  Employed full-time   
4  The quality of OSS and closed source software ...  Employed full-time   

          Country Student                                            EdLevel  \
0   United States      No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
1     New Zealand      No  Some college/university study without earning ...   
2   United States      No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
3  United Kingdom      No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
4       Australia      No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                      UndergradMajor  ...  \
0  Computer science, computer engineering, or sof...  ...   
1  Computer science, computer engineering, or sof...  ...   
2  Computer science, computer engineering, or sof...  ...   
3                                                NaN  ...   
4  Computer science, computer engineering, or sof...  ...   

                                        SONewContent   Age Gender Trans  \
0  Tech articles written by other developers;Indu...  22.0    Man    No   
1                                                NaN  23.0    Man    No   
2  Tech articles written by other developers;Cour...  28.0    Man    No   
3  Tech articles written by other developers;Indu...  26.0    Man    No   
4  Tech articles written by other developers;Indu...  29.0    Man    No   

                 Sexuality                              Ethnicity Dependents  \
0  Straight / Heterosexual           White or of European descent         No   
1                 Bisexual           White or of European descent         No   
2  Straight / Heterosexual           White or of European descent        Yes   
3  Straight / Heterosexual           White or of European descent         No   
4  Straight / Heterosexual  Hispanic or Latino/Latina;Multiracial         No   

            SurveyLength                  SurveyEase  \
0  Appropriate in length                        Easy   
1  Appropriate in length  Neither easy nor difficult   
2  Appropriate in length                        Easy   
3  Appropriate in length  Neither easy nor difficult   
4  Appropriate in length                        Easy   

  NormalizedAnnualCompensation  
0                      61000.0  
1                     138000.0  
2                      90000.0  
3                     348000.0  
4                      90000.0  

[5 rows x 86 columns]